### ISSUE TITLE
[react/jsx-no-bind] ignore function in props using args of iterator function

### ISSUE THREAD

**Issue by SyMind** [yannickcr/eslint-plugin-react#3013]
Should never report warning in follow example, but now warning with message `JSX props should not use arrow functions`.

```javascript
const Foo = ({list}) => (
    <div>
        {
            list.map(item => (
                <div
                    key={item.id}
                    onClick={() => alert(item.text)} // JSX props should not use arrow functions
                >
                    click me
                </div>
            ))
        }
    </div>
);
```

**Comment by ljharb** [comment]
@SyMind why should that not report a warning?

**Comment by SyMind** [comment]
I think, if function in props use args of iterator function, it's hard to avoid using a new function reference in every in each iterator. In my case above, this can not be done with `useCallback`, because in each iterator `alert` use `item.text`.

**Comment by ljharb** [comment]
"it's hard" is irrelevant - it's possible (and in this case, not all that hard):

```jsx
function ListItem({ id, text }) {
    const callback = useCallback(() => alert(item.text), item.text);
    return (
        <div onClick={callback}>
            click me
        </div>
    );
}

function Foo({list}) (
    return (
        <div>
            {list.map(item => <ListItem key={item.id} {...item} />)}
        </div>
    );
);
```
